design-documents/atomics-ABI.rst

```diff
@@ -0,0 +1,308 @@
+  +=============+==============================================================+=============================================================================+
+  | ARMARM_     | DDI 0487                                                     | Arm Architecture Reference Manual Armv8 for Armv8-A architecture profile    |
+  +-------------+--------------------------------------------------------------+-----------------------------------------------------------------------------+
+  | PAPER_      | CGO paper                                                    | Compiler Testing with Relaxed Memory Models                                 |
+  +-------------+--------------------------------------------------------------+-----------------------------------------------------------------------------+
+
+
+
+Note: At the time of writing C23 is not released, as such ISO C17 is considered
+the latest published document.
+
+Use-cases known of so far
+-------------------------
+
+
+A Baseline: Describing current implementations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ABI we provide is a baseline specification that compilers should or do implement.
+The ABI provides a grounds to be compatible across all versions of the Armv8 architecture. Most
+of the mappings in the ABI are already implemented in LLVM and GCC and this ABI ratifies
+a decade of established practice, and provides alternatives where the current practice
+is incompatible.
+
+
+Sub-ABIs and ABI-islands: Departing from the baseline (or 'mainland')
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We do *not* require that compilers implement this ABI. Implementers can specify their own
+ABI, whether it is a subset of the allowed mappings of the baseline ABI (a sub-ABI), or 
+uses different mappings altogether (an ABI-island). Currently, sub-ABIs and ABI-islands implicitly
+arise with each new architecture release, and implementers quickly find new candidate mappings
+that are performant on their machines. Such mappings are proposed or added to mainstream
+compilers. However due to the lack of a baseline specification or widespread
+concurrency expertise, testing such mappings has been a challenge and concurrency bugs have been
+unintentionally introduced into compilers when new mappings are added.
+
+We need a baseline ABI in order to determine if a given sub-ABI respects or departs
+from the baseline. Adding command-line options is a logical consequence of defining such an ABI, 
+and makes it possible to track ABI compatibility of concurrent programs at compile or link-time,
+rather than runtime. It is the responsibility of the sub-ABI maintainer to ensure code built
+under their ABI does not mix with code built under the baseline. But a baseline must exist, 
+for sub-ABI compatibility to be decided in the first place.
+
+A baseline provides the means to describe or contain ABI-islands. Where a compiler implementation
```

Comment:
I'd delete the first sentence. It uses ABI-Islands before introducing the concept.